--- linux-2.6.36/scripts/mod/modpost.c.orig	2010-10-20 22:30:22.000000000 +0200
+++ linux-2.6.36/scripts/mod/modpost.c	2010-10-25 12:22:49.053556168 +0200
@@ -16,7 +16,8 @@
 #include <ctype.h>
 #include <string.h>
 #include "modpost.h"
-#include "../../include/generated/autoconf.h"
+// PLD architectures don't use CONFIG_SYMBOL_PREFIX
+//#include "../../include/generated/autoconf.h"
 #include "../../include/linux/license.h"
 
 /* Some toolchains use a `_' prefix for all user symbols. */
